std/digest/murmurhash.d

```diff
@@ -496,29 +499,81 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
-        buffer.data[bufferSize .. $] = data[0 .. bufferLeeway];
-        putElement(buffer.block);
-        data = data[bufferLeeway .. $];
-
-        // Do main work: process chunks of `Element.sizeof` bytes.
-        const numElements = data.length / Element.sizeof;
-        const remainderStart = numElements * Element.sizeof;
-        foreach (ref const Element block; cast(const(Element[]))(data[0 .. remainderStart]))
+
+        // Check if we have some leftover data in the buffer. Then fill the first block buffer.
+        const bool haveLeewayElement = (bufferSize != 0);
+        if (haveLeewayElement)
         {
-            putElement(block);
+            const bufferLeeway = Element.sizeof - bufferSize;
+            assert(bufferLeeway < Element.sizeof);
+            buffer.data[bufferSize .. $] = data[0 .. bufferLeeway];
+            putElement(buffer.block);
+            data = data[bufferLeeway .. $];
+        }
+
+        size_t numChunks = 0;
+        version(haveUnalignedLoads)
+        {
+            // Do main work: process chunks of `Element.sizeof` bytes.
+            numChunks = data.length / Element.sizeof;
+            foreach (ref const Element block; cast(const(Element[]))(data[0 .. numChunks * Element.sizeof]))
+            {
+                putElement(block);
+            }
+        }
+        else
+        {
+            size_t processChunks(TChunk)()
+            {
+                const numFullChunks = data.length / TChunk.sizeof;
+                foreach (ref const chunk; cast(const(TChunk)[]) data[0 .. numFullChunks * TChunk.sizeof])
+                {
+                    const block = cast(Element) chunk;
+                    putElement(block);
+                }
+                return numFullChunks;
+            }
+
+            const startAddress = cast(size_t) data.ptr;
+            static if (size >= 64)
+            {
+                if ((startAddress & 7) == 0)
+                    numChunks = processChunks!(ulong[size / 64])();
+                else if ((startAddress & 3) == 0)
+                    numChunks = processChunks!(uint[size / 32])();
+                else if ((startAddress & 1) == 0)
+                    numChunks = processChunks!(ushort[size / 16])();
+                else
+                    numChunks = processChunks!(ubyte[size / 8])();
+            }
+            else
+            {
+                static assert(size == 32);
+                if ((startAddress & 3) == 0)
+                    numChunks = processChunks!(Element)();
+                else
+                {
+                    numChunks = data.length / Element.sizeof;
+                    foreach (ref const chunk; cast(const(ubyte[4])[]) data[0 .. numChunks * Element.sizeof])
+                    {
+                        import core.stdc.string : memcpy;
+                        Element block;
+                        ()@trusted{memcpy(&block, &chunk, Element.sizeof);}();
+                        putElement(block);
```

Comment:
Merged and reorganized the 32-bit alignment fix I added yesterday, @kinke, here's the `memcpy` workaround David suggested earlier.  I benchmarked it on unaligned data using [@gchatelet's benchmark](https://github.com/gchatelet/murmurhash3_d/blob/master/benchmark.d), it's usually only 1-2% slower, around 4 GiB/s.